pkg/util/utils.go

```diff
@@ -723,3 +726,304 @@ func SocketPath() (string, error) {
 	// Glue the socket path together
 	return filepath.Join(xdg, "podman", "podman.sock"), nil
 }
+
+// ImageScpOptions provide options for securely copying images to and from a remote host
+type ImageScpOptions struct {
+	// Remote determines if this entity is operating on a remote machine
+	Remote bool `json:"remote,omitempty"`
+	// File is the input/output file for the save and load Operation
+	File string `json:"file,omitempty"`
+	// Quiet Determines if the save and load operation will be done quietly
+	Quiet bool `json:"quiet,omitempty"`
+	// Image is the image the user is providing to save and load
+	Image string `json:"image,omitempty"`
+	// Transfer determines whether we are transferring an image locally
+	Transfer bool `json:"transfer,omitempty"`
+	// User is used in conjunction with Transfer to determine if a valid user was given to save from/load into
+	User string `json:"user,omitempty"`
+}
+
+// ParseImageSCPArgs returns the valid connection data based off of the information provided by the user
+// args is an array of the command arguments and cfg is tooling configuration used to get service destinations
+// returned is serv and an error if applicable. serv is a map of service destinations with the connection name as the index
+// this connection name is intended to be used as EngineConfig.ServiceDestinations
+// this function modifies the global scpOpt entities: FromRemote, ToRemote, Connections, and SourceImageName
+func ParseImageSCPArgs(args []string) (*ImageScpOptions, *ImageScpOptions, []string, error) {
+	source := ImageScpOptions{}
+	dest := ImageScpOptions{}
+
+	cliConnections := []string{}
+	switch len(args) {
+	case 1:
+		if strings.Contains(args[0], "@localhost") { // image transfer between users
+			source.User = strings.Split(args[0], "@")[0]
+			source.Transfer = true
+			if len(strings.Split(args[0], "::")) > 1 {
+				source.Image = strings.Split(args[0], "::")[1]
+			} else {
+				return nil, nil, nil, errors.New("No source image provided")
+			}
+		} else if strings.Contains(args[0], "::") {
+			source.Image = strings.Split(args[0], "::")[1] // this will get checked/set again once we validate connections
+			source.Remote = true
+			cliConnections = append(cliConnections, args[0])
+		} else {
+			err := ValidateImageName(args[0])
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			source.Image = args[0]
+		}
+	case 2:
+		if strings.Contains(args[0], "@localhost") || strings.Contains(args[1], "@localhost") { // image transfer between users
+			source.Transfer = true
+			if len(strings.Split(args[0], "::")) > 1 && RemoteArgLength(args[0], 1) > 0 && len(strings.Split(args[1], "::")) > 1 && RemoteArgLength(args[1], 1) > 0 { // we cannot specify a new image name
```

Comment:
Code like this almost always screams for a swich statement.

Reply:
Or multiple switch statements.

Reply:
Ok, I'll throw in a switch in this section

Reply:
Actually multiple switches would make it easier to read in my opinion. Switch on first half of option and then either subswich or if/then else, if only binary subtest.

Reply:
what about this @rhatdan...

```
for i, arg := range args { // iterate over both of our arguments, the same logic applies to both
		switch RemoteArgLength(arg, 1) {
		case 0:
			if len(strings.Split(arg, "::")) > 1 { // dest_user@localhost:: image
				dest.User = strings.Split(arg, "@")[0]
				source.Image = args[len(args)-i]
			}
		default:
			if len(strings.Split(arg, "::")) > 1 { // first argument contains image and therefore is our user
				source.User = strings.Split(arg, "@")[0]
				dest.User = strings.Split(args[len(args)-i], "@")[0]
				source.Image = strings.Split(arg, "::")[1]
			}

		}
}
```

This is the same logic. The checks are too different to make an easy switch for each case, so if I make a switch for each argument that should work. We do the same exact checks for arg 0 and arg 1

Reply:
so the full snipped would be... 

```
case 2:
if strings.Contains(args[0], "@localhost") || strings.Contains(args[1], "@localhost") { // image transfer between users
	source.Transfer = true
	if len(strings.Split(args[0], "::")) > 1 && RemoteArgLength(args[0], 1) > 0 && len(strings.Split(args[1], "::")) > 1 && RemoteArgLength(args[1], 1) > 0 { // we cannot specify a new image name
		return nil, nil, nil, errors.Wrapf(define.ErrInvalidArg, "Image scp does not support image renaming. Please do not specify a destination name")
	}
	for i, arg := range args { // iterate over both of our arguments, the same logic applies to both
		switch RemoteArgLength(arg, 1) {
		case 0:
			if len(strings.Split(arg, "::")) > 1 { // this argument contains no image
				dest.User = strings.Split(arg, "@")[0]
				source.Image = args[len(args)-i]
			}
		default:
			if len(strings.Split(arg, "::")) > 1 { // this argument contains an image
				source.User = strings.Split(arg, "@")[0]
				dest.User = strings.Split(args[len(args)-i], "@")[0]
				source.Image = strings.Split(arg, "::")[1]
			}

		}
	}
	if len(source.Image) == 0 {
				return nil, nil, nil, errors.New("No source image provided. Please either specify a local or remote image to pass between hosts")
}
```

Reply:
Looks great